sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -1278,9 +1407,27 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     case DateType =>
       (c, evPrim, evNull) => code"$evNull = true;"
     case TimestampType =>
-      (c, evPrim, evNull) => code"$evPrim = (long) ${timestampToIntegerCode(c)};"
+      (c, evPrim, evNull) => code"$evPrim = (long) ${timestampToLongCode(c)};"
+    case DecimalType() if failOnIntegerOverflow =>
+      (c, evPrim, evNull) =>
+        code"""
+          try {
+            $evPrim = $c.toJavaBigInteger().longValueExact();
```

Comment:
this would mean creating several java objects and I don't think it is going to have good perf. What about using `toLong` with a flag or create a `toLongExact`?